### PR TITLE
Fix issues with world limit

### PR DESCRIPTION
**Pull request by ImMorpheus**
Fix https://github.com/SpongePowered/SpongeCommon/issues/168
Fix https://github.com/SpongePowered/SpongeCommon/issues/651
Supersedes https://github.com/SpongePowered/SpongeCommon/pull/1612

### CODE REVIEW

**Comment by ImMorpheus** [comment]
~qa

**Comment by gabizou** [comment]
I see you've changed it for biomes, but what about for block retrieval?

**Comment by ImMorpheus** [comment]
Test code:

See test plugin.

Also:
```
((Player) src).transferToWorld(((Player) src).getWorld(), Vector3d.from(Double.MAX_VALUE, ((Player) src).getPosition().getY(), Double.MAX_VALUE));
```
is working _"as intended"_.

Results:

/test 30000000 20 30000000

```
org.spongepowered.api.util.PositionOutOfBoundsException: Position is out of bounds: expected in range (-30000000, 0, -30000000) to (29999999, 255, 29999999) but got (30000000, 20, 30000000)
	at net.minecraft.world.WorldServer.checkBlockBounds(WorldServer.java:2593) ~[WorldServer.class:?]
	at net.minecraft.world.WorldServer.setBlock(WorldServer.java:2573) ~[WorldServer.class:?]
	at net.minecraft.world.World.setBlock(World.java:5847) ~[World.class:?]
	at org.spongepowered.test.DevTest.lambda$onInit$0(DevTest.java:33) ~[DevTest.class:?]
	at org.spongepowered.api.command.spec.CommandSpec.process(CommandSpec.java:388) ~[CommandSpec.class:?]
	at org.spongepowered.api.command.dispatcher.SimpleDispatcher.process(SimpleDispatcher.java:341) ~[SimpleDispatcher.class:?]
	at org.spongepowered.common.command.SpongeCommandManager.process(SpongeCommandManager.java:328) [SpongeCommandManager.class:?]
	at net.minecraft.command.ServerCommandManager.executeCommand(ServerCommandManager.java:1083) [ServerCommandManager.class:?]
	at net.minecraft.network.NetHandlerPlayServer.handleSlashCommand(NetHandlerPlayServer.java:1001) [NetHandlerPlayServer.class:?]
	at net.minecraft.network.NetHandlerPlayServer.processChatMessage(NetHandlerPlayServer.java:977) [NetHandlerPlayServer.class:?]
	at net.minecraft.network.play.client.CPacketChatMessage.processPacket(CPacketChatMessage.java:47) [CPacketChatMessage.class:?]
	at net.minecraft.network.play.client.CPacketChatMessage.processPacket(CPacketChatMessage.java:8) [CPacketChatMessage.class:?]
	at org.spongepowered.common.network.PacketUtil.onProcessPacket(PacketUtil.java:148) [PacketUtil.class:?]
	at net.minecraft.network.PacketThreadUtil$1.redirect$onProcessPacket$zla000(PacketThreadUtil.java:539) [PacketThreadUtil$1.class:?]
	at net.minecraft.network.PacketThreadUtil$1.run(PacketThreadUtil.java:21) [PacketThreadUtil$1.class:?]
	at java.util.concurrent.Executors$RunnableAdapter.call$$$capture(Executors.java:511) [?:1.8.0_171]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java) [?:1.8.0_171]
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266) [?:1.8.0_171]
	at java.util.concurrent.FutureTask.run(FutureTask.java) [?:1.8.0_171]
	at net.minecraft.util.Util.runTask(Util.java:53) [Util.class:?]
	at org.spongepowered.common.SpongeImplHooks.onUtilRunTask(SpongeImplHooks.java:288) [SpongeImplHooks.class:?]
	at net.minecraft.server.MinecraftServer.redirect$onRun$zjf000(MinecraftServer.java:4169) [MinecraftServer.class:?]
	at net.minecraft.server.MinecraftServer.updateTimeLightAndEntities(MinecraftServer.java:798) [MinecraftServer.class:?]
	at net.minecraft.server.dedicated.DedicatedServer.updateTimeLightAndEntities(DedicatedServer.java:414) [DedicatedServer.class:?]
	at net.minecraft.server.MinecraftServer.tick(MinecraftServer.java:743) [MinecraftServer.class:?]
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:592) [MinecraftServer.class:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_171]
```

/test1 30000000 66 30000000

`[Server thread/INFO] [STDOUT]: [org.spongepowered.test.DevTest:lambda$onInit$1:51]: minecraft:air`
(should be redsand)
(fun fact: /sponge blockinfo is working fine and shows the correct block)

Something strange:

/setblock 30000000 90 30000000 glowstone

"The block couldn't be placed."

But that command is working fine on vanilla mc. 

I'm working on it.

**Comment by gabizou** [comment]
Word of recommendation, make a test plugin to commit to it so it's easier when maintainers get to test the PR as plugin usage code is already provided ;)

**Comment by phit** [comment]
name the commands something meaningful, so they dont conflict when you have all test plugins loaded
can't name them all test :P

File: testplugins/src/main/java/org/spongepowered/test/WorldLimitTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                            if (src instanceof Player) {
+                                Vector3d coord = args.<Vector3d>getOne("coord").get();
+
+                                ((Player) src).getWorld().setBlock(coord.getFloorX(), coord.getFloorY(), coord.getFloorZ(), BlockState.builder().blockType(BlockTypes.STONE).build());
+                            }
+                            return CommandResult.success();
+                        })
+                        .arguments(GenericArguments.vector3d(Text.of("coord")))
+                        .build(),
+                "test"
```
Comment (ImMorpheus) [inline testplugins/src/main/java/org/spongepowered/test/WorldLimitTest.java]: //TODO find a better name

EDIT: Damn, phit beat me to it.

**Comment by ImMorpheus** [comment]
Not sure what to do about setBlock, setBiome, etc.

Should plugins be allowed to place blocks or modifying the world beyond the world limit ?

**Comment by gabizou** [comment]
> Should plugins be allowed to place blocks or modifying the world beyond the world limit ?

Does the engine support it? What happens when you use the command `/setblock` in an out of bounds area?

Tested locally with forge:


Looks like the engine doesn't crash but tells the player it doesn't work.. I'd tell plugins that it simply failed (with the method already having a "valid" return type).

**Comment by ImMorpheus** [comment]
> Does the engine support it? What happens when you use the command /setblock in an out of bounds area?
> Looks like the engine doesn't crash but tells the player it doesn't work.. 

Interesting. Looks like the problem is the edge.
`/setblock 30000000 70 30000000 glowstone`

Vanilla: Everything works.
Sponge: The block couldn't be placed.

**Comment by ImMorpheus** [comment]
Fixed.

Squashed and rebased.

**Comment by ImMorpheus** [comment]
Status ? This could be included in the RB since it's crash-related.

**Pull request merged**